miniapps/dpg/util/complexweakform.hpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_COMPLEX_DPGWEAKFORM
+#define MFEM_COMPLEX_DPGWEAKFORM
+
+#include "mfem.hpp"
+#include "complexstaticcond.hpp"
+
+namespace mfem
+{
+
+/** @brief Class representing the complex DPG weak formulation. */
+class ComplexDPGWeakForm
```

Comment:
Is this worh upstreaming to the MFEM library?

Reply:
I am open to doing this. Whatever the team decides I am okay with it.

Reply:
I am fine with moving this to the library. It may be more practical to have one class `DGPWeakForm` that includes complex support. There is a lot of similarity between `DGPWeakForm` and `ComplexDGPWeakForm`, so that any change in one has to be replicated in the other. This makes the code harder to maintain. Two reasons to keep them separate are (i) to avoid having lots of complicated switches between real and complex cases, and (ii) to avoid unnecessary computation in the real case. I don't see much complication in combining the classes, with a real/complex flag. Functions could be overloaded for the real case, with an assertion of the real flag. The overhead of imaginary data pointers is negligible.

Reply:
We kept Complex support separated everywhere else so far, so I vote for keeping it separated here as well